File: apps/web/src/app/api/assets/_lib/localAssetStore.ts
import { randomUUID } from 'node:crypto'
import { mkdir, readFile, writeFile } from 'node:fs/promises'
import path from 'node:path'
import type {
  TangentAssetDataUrlInput,
  TangentAssetRecord,
  TangentAssetThumbnailInput,
} from '@/features/assets/assetTypes'
import type { ApiRequestContext } from '../../_lib/apiRequestContext'

const allowedMimeTypes = new Set(['image/jpeg', 'image/png', 'image/webp'])
const maxAssetBytes = 100 * 1024 * 1024
const storageRoot = process.env.TANGENT_ASSET_STORAGE_DIR ?? getDefaultLocalStorageRoot('.tangent-assets')
const assetsRoot = path.join(storageRoot, 'assets')

type ParsedDataUrl = {
  buffer: Buffer
  mime: string
}

export async function createLocalAssetFromDataUrl(input: TangentAssetDataUrlInput, context: ApiRequestContext) {
  const original = parseImageDataUrl(input.dataUrl)
  assertImageMime(original.mime)
  assertAssetSize(original.buffer.byteLength)

  const assetId = `asset_${randomUUID()}`
  const assetDir = path.join(assetsRoot, assetId)
  await mkdir(assetDir, { recursive: true })

  const originalFileName = `original.${getExtension(original.mime)}`
  await writeFile(path.join(assetDir, originalFileName), original.buffer)

  const thumbnailUrls: Partial<Pick<
    TangentAssetRecord,
    'thumbnail1024Url' | 'thumbnail256Url' | 'thumbnail512Url'
  >> = {}
  await writeThumbnail(assetDir, assetId, 256, input.thumbnails?.[256], thumbnailUrls)
  await writeThumbnail(assetDir, assetId, 512, input.thumbnails?.[512], thumbnailUrls)
  await writeThumbnail(assetDir, assetId, 1024, input.thumbnails?.[1024], thumbnailUrls)

  const record: TangentAssetRecord = {
    ...thumbnailUrls,
    byteSize: original.buffer.byteLength,
    createdAt: new Date().toISOString(),
    createdBy: context.userId,
    height: input.height,
    id: assetId,
    mime: original.mime,
    origin: input.origin,
    originalUrl: createFileUrl(assetId, originalFileName),
    storage: 'local-dev',
    title: input.title || input.fileName || 'Image',
    width: input.width,
    workspaceId: context.workspaceId,
  }
  await writeAssetRecord(assetDir, record)
  return record
}

export async function createLocalAssetFromUpload(input: {
  bytes: ArrayBuffer
  fileName?: string
  height?: number
  mime: string
  origin?: TangentAssetRecord['origin']
  title?: string
  width?: number
}, context: ApiRequestContext) {
  assertImageMime(input.mime)
  const assetId = `asset_${randomUUID()}`
  const assetDir = path.join(assetsRoot, assetId)
  await mkdir(assetDir, { recursive: true })

  const originalFileName = `original.${getExtension(input.mime)}`
  const buffer = Buffer.from(input.bytes)
  assertAssetSize(buffer.byteLength)
  await writeFile(path.join(assetDir, originalFileName), buffer)

  const record: TangentAssetRecord = {
    byteSize: buffer.byteLength,
    createdAt: new Date().toISOString(),
    createdBy: context.userId,
    height: input.height ?? 0,
    id: assetId,
    mime: input.mime,
    origin: input.origin ?? 'upload',
    originalUrl: createFileUrl(assetId, originalFileName),
    storage: 'local-dev',
    title: input.title || input.fileName || 'Image',
    width: input.width ?? 0,
    workspaceId: context.workspaceId,
  }
  await writeAssetRecord(assetDir, record)
  return record
}

export async function getLocalAssetRecord(assetId: string, context: ApiRequestContext) {
  const record = await readStoredAssetRecord(assetId, context)
  assertWorkspaceAccess(record, context)
  return record
}

export async function readLocalAssetFile(assetId: string, fileName: string, context: ApiRequestContext) {
  assertSafePathSegment(assetId)
  assertSafePathSegment(fileName)
  if (!context.isDevFallback) {
    await getLocalAssetRecord(assetId, context)
  }
  const filePath = path.join(assetsRoot, assetId, fileName)
  const file = await readFile(filePath)
  const body = file.buffer.slice(file.byteOffset, file.byteOffset + file.byteLength) as ArrayBuffer
  return { file: body, mime: getMimeFromFileName(fileName) }
}

async function readStoredAssetRecord(assetId: string, context: ApiRequestContext) {
  assertSafePathSegment(assetId)
  const recordPath = path.join(assetsRoot, assetId, 'metadata.json')
  const raw = await readFile(recordPath, 'utf8')
  return normalizeAssetRecord(JSON.parse(raw) as Partial<TangentAssetRecord>, context)
}

function parseImageDataUrl(dataUrl: string): ParsedDataUrl {
  const match = /^data:([^;,]+);base64,([a-zA-Z0-9+/=\s]+)$/s.exec(dataUrl)
  if (!match) throw new Error('Invalid image data URL.')
  const mime = match[1]?.toLowerCase() ?? ''
  assertImageMime(mime)
  const base64 = (match[2] ?? '').replace(/\s+/g, '')
  assertAssetSize(estimateBase64ByteLength(base64))
  const buffer = Buffer.from(base64, 'base64')
  assertAssetSize(buffer.byteLength)
  return {
    buffer,
    mime,
  }
}

async function writeThumbnail(
  assetDir: string,
  assetId: string,
  size: 256 | 512 | 1024,
  thumbnail: TangentAssetThumbnailInput | undefined,
  urls: Partial<Pick<TangentAssetRecord, 'thumbnail1024Url' | 'thumbnail256Url' | 'thumbnail512Url'>>
) {
  if (!thumbnail?.dataUrl) return
  const parsed = parseImageDataUrl(thumbnail.dataUrl)
  assertImageMime(parsed.mime)
  assertAssetSize(parsed.buffer.byteLength)
  const fileName = `thumb-${size}.${getExtension(parsed.mime)}`
  await writeFile(path.join(assetDir, fileName), parsed.buffer)
  urls[`thumbnail${size}Url` as keyof typeof urls] = createFileUrl(assetId, fileName)
}

async function writeAssetRecord(assetDir: string, record: TangentAssetRecord) {
  await writeFile(path.join(assetDir, 'metadata.json'), `${JSON.stringify(record, null, 2)}\n`)
}

function createFileUrl(assetId: string, fileName: string) {
  return `/api/assets/files/${assetId}/${fileName}`
}

function assertImageMime(mime: string) {
  if (!allowedMimeTypes.has(mime)) throw new Error('Unsupported image MIME type.')
}

function assertAssetSize(byteSize: number) {
  if (byteSize > maxAssetBytes) throw new Error('Image must be 100MB or smaller.')
}

function estimateBase64ByteLength(base64: string) {
  if (!base64 || base64.length % 4 !== 0) throw new Error('Invalid image data URL.')
  const padding = base64.endsWith('==') ? 2 : base64.endsWith('=') ? 1 : 0
  return Math.floor((base64.length * 3) / 4) - padding
}

function getExtension(mime: string) {
  if (mime === 'image/png') return 'png'
  if (mime === 'image/webp') return 'webp'
  return 'jpg'
}

function getMimeFromFileName(fileName: string) {
  if (fileName.endsWith('.png')) return 'image/png'
  if (fileName.endsWith('.webp')) return 'image/webp'
  return 'image/jpeg'
}

function getDefaultLocalStorageRoot(directoryName: string) {
  return path.join(/*turbopackIgnore: true*/ process.cwd(), '..', '..', directoryName)
}

function normalizeAssetRecord(record: Partial<TangentAssetRecord>, context: ApiRequestContext): TangentAssetRecord {
  return {
    byteSize: record.byteSize ?? 0,
    createdAt: record.createdAt ?? new Date(0).toISOString(),
    createdBy: record.createdBy ?? context.userId,
    height: record.height ?? 0,
    id: record.id ?? '',
    mime: record.mime ?? 'image/jpeg',
    origin: record.origin ?? 'upload',
    originalUrl: record.originalUrl ?? '',
    storage: record.storage ?? 'local-dev',
    thumbnail1024Url: record.thumbnail1024Url,
    thumbnail256Url: record.thumbnail256Url,
    thumbnail512Url: record.thumbnail512Url,
    title: record.title ?? 'Image',
    width: record.width ?? 0,
    workspaceId: record.workspaceId ?? context.workspaceId,
  }
}

function assertWorkspaceAccess(record: TangentAssetRecord, context: ApiRequestContext) {
  if (process.env.NODE_ENV !== 'production' && process.env.TANGENT_REQUIRE_API_AUTH !== '1') return
  if (record.workspaceId !== context.workspaceId) {
    throw new Error('Asset not found in workspace.')
  }
}

function assertSafePathSegment(value: string) {
  if (!/^[a-zA-Z0-9._-]+$/.test(value) || value.includes('..')) {
    throw new Error('Invalid asset path.')
  }
}

File: services/api/tangent_api/storage/local_asset_store.py
import json
import os
from datetime import datetime, timezone
from pathlib import Path
from typing import Optional
from uuid import uuid4

from fastapi import HTTPException, UploadFile

from tangent_api.request_context import ApiRequestContext
from tangent_api.schemas import AssetDataUrlRequest, AssetRecord, AssetThumbnailInput
from tangent_api.storage.asset_store_common import (
    assert_asset_size,
    assert_image_mime,
    assert_safe_path_segment,
    assert_workspace_access,
    extension_for_mime,
    file_url,
    parse_image_data_url,
)


def create_asset_from_data_url(
    input_data: AssetDataUrlRequest,
    context: ApiRequestContext,
) -> AssetRecord:
    original = parse_image_data_url(input_data.data_url)
    assert_image_mime(original.mime)
    assert_asset_size(len(original.content))

    asset_id = f"asset_{uuid4()}"
    asset_dir = _asset_dir(asset_id)
    asset_dir.mkdir(parents=True, exist_ok=True)

    original_file_name = f"original.{extension_for_mime(original.mime)}"
    (asset_dir / original_file_name).write_bytes(original.content)

    thumbnail_urls = _write_thumbnails(asset_dir, asset_id, input_data.thumbnails)
    record = AssetRecord(
        **thumbnail_urls,
        byteSize=len(original.content),
        createdAt=datetime.now(timezone.utc).isoformat(),
        createdBy=context.user_id,
        height=input_data.height,
        id=asset_id,
        mime=original.mime,
        origin=input_data.origin,
        originalUrl=file_url(asset_id, original_file_name),
        storage="local-dev",
        title=input_data.title or input_data.file_name or "Image",
        width=input_data.width,
        workspaceId=context.workspace_id,
    )
    _write_asset_record(asset_dir, record)
    return record


async def create_asset_from_upload(
    file: UploadFile,
    context: ApiRequestContext,
    origin: str,
    title: Optional[str],
    width: int,
    height: int,
) -> AssetRecord:
    mime = file.content_type or ""
    assert_image_mime(mime)
    content = await file.read()
    assert_asset_size(len(content))

    asset_id = f"asset_{uuid4()}"
    asset_dir = _asset_dir(asset_id)
    asset_dir.mkdir(parents=True, exist_ok=True)

    original_file_name = f"original.{extension_for_mime(mime)}"
    (asset_dir / original_file_name).write_bytes(content)

    record = AssetRecord(
        byteSize=len(content),
        createdAt=datetime.now(timezone.utc).isoformat(),
        createdBy=context.user_id,
        height=height,
        id=asset_id,
        mime=mime,
        origin=origin,
        originalUrl=file_url(asset_id, original_file_name),
        storage="local-dev",
        title=title or file.filename or "Image",
        width=width,
        workspaceId=context.workspace_id,
    )
    _write_asset_record(asset_dir, record)
    return record


def create_asset_from_bytes(
    content: bytes,
    mime: str,
    context: ApiRequestContext,
    origin: str,
    title: Optional[str],
    width: int,
    height: int,
) -> AssetRecord:
    assert_image_mime(mime)
    assert_asset_size(len(content))

    asset_id = f"asset_{uuid4()}"
    asset_dir = _asset_dir(asset_id)
    asset_dir.mkdir(parents=True, exist_ok=True)

    original_file_name = f"original.{extension_for_mime(mime)}"
    (asset_dir / original_file_name).write_bytes(content)

    record = AssetRecord(
        byteSize=len(content),
        createdAt=datetime.now(timezone.utc).isoformat(),
        createdBy=context.user_id,
        height=height,
        id=asset_id,
        mime=mime,
        origin=origin,
        originalUrl=file_url(asset_id, original_file_name),
        storage="local-dev",
        title=title or "Image",
        width=width,
        workspaceId=context.workspace_id,
    )
    _write_asset_record(asset_dir, record)
    return record


def get_asset_record(asset_id: str, context: ApiRequestContext) -> AssetRecord:
    assert_safe_path_segment(asset_id)
    path = _asset_dir(asset_id) / "metadata.json"
    if not path.exists():
        raise HTTPException(status_code=404, detail="Asset not found.")
    record = AssetRecord.model_validate(json.loads(path.read_text(encoding="utf-8")))
    assert_workspace_access(record, context)
    return record


def get_asset_file_path(asset_id: str, file_name: str, context: ApiRequestContext) -> Path:
    assert_safe_path_segment(asset_id)
    assert_safe_path_segment(file_name)
    get_asset_record(asset_id, context)
    path = _asset_dir(asset_id) / file_name
    if not path.exists():
        raise HTTPException(status_code=404, detail="Asset file not found.")
    return path


def _write_thumbnails(
    asset_dir: Path,
    asset_id: str,
    thumbnails: Optional[dict[int, AssetThumbnailInput]],
) -> dict[str, str]:
    urls: dict[str, str] = {}
    for size in (256, 512, 1024):
        thumbnail = thumbnails.get(size) if thumbnails else None
        if not thumbnail:
            continue
        parsed = parse_image_data_url(thumbnail.data_url)
        assert_image_mime(parsed.mime)
        assert_asset_size(len(parsed.content))
        file_name = f"thumb-{size}.{extension_for_mime(parsed.mime)}"
        (asset_dir / file_name).write_bytes(parsed.content)
        urls[f"thumbnail{size}Url"] = file_url(asset_id, file_name)
    return urls


def _write_asset_record(asset_dir: Path, record: AssetRecord) -> None:
    asset_dir.joinpath("metadata.json").write_text(
        json.dumps(record.model_dump(by_alias=True), ensure_ascii=False, indent=2) + "\n",
        encoding="utf-8",
    )


def _storage_root() -> Path:
    return Path(os.getenv("TANGENT_ASSET_STORAGE_DIR", ".tangent-assets"))


def _asset_dir(asset_id: str) -> Path:
    return _storage_root() / "assets" / asset_id
